Storage subsystem version 2

shamil-gadelshin opened the pull request:
This PR contains rewritten from scratch version 2 of the Storage subsystem. The previous version (pallets `storage` and `service discovery`) was removed.  The data objects migration for the `content directory` was removed as well.

## Initial spec links
- [spec](https://github.com/Joystream/joystream/issues/2224)
- [utilization model](https://github.com/Joystream/joystream/issues/2359)

## Notable abstractions
- DataObjectStorage - a trait for the pallet integration. Exposes pallet public methods
- ModuleAccount - a trait for the pallet specific accounts
- BagManager - a facade over dynamic and static bag operations. Provides a common interface.
- StorageBucketPicker - implements random (and complementary sequential) algorithm for picking storage buckets on dynamic bag creation

## Pallet functionality could be split into three distinct groups
- extrinsics for the working group leader
- extrinsics for the storage provider
- public methods for the pallet integration

#### Working group leader extrinsics
- create_storage_bucket
- update_storage_buckets_for_bag 
- delete_storage_bucket
- invite_storage_bucket_operator
- cancel_storage_bucket_operator_invite
- remove_storage_bucket_operator
- update_uploading_blocked_status
- update_data_size_fee
- update_storage_buckets_per_bag_limit
- update_storage_buckets_voucher_max_limits
- update_number_of_storage_buckets_in_dynamic_bag_creation_policy
- update_blacklist

#### Storage provider extrinsics
- accept_storage_bucket_invitation
- set_storage_operator_metadata
- set_storage_bucket_voucher_limits
- update_storage_bucket_status
- accept_pending_data_objects

#### Public integration methods are exposed via the DataObjectStorage.
- can_upload_data_objects
- upload_data_objects
- can_move_data_objects
- move_data_objects
- can_delete_data_objects
- delete_data_objects
- can_delete_dynamic_bag
- delete_dynamic_bag
- can_create_dynamic_bag
- create_dynamic_bag

#### Pallet constants
- MaxStorageBucketNumber
- MaxNumberOfDataObjectsPerBag
- DataObjectDeletionPrize
- BlacklistSizeLimit
- StorageBucketsPerBagValueConstraint
- DefaultMemberDynamicBagCreationPolicy
- DefaultChannelDynamicBagCreationPolicy


## Next iteration tasks
1. adding authentication_key.
2. adjust storage pallet-related constants in runtime.
3. consider adding config() to enable future storage migration.
4. make public methods as root extrinsics to enable storage-node dev mode.
5. consider adding a counter for the total objects size for a bag.

## Far future [changes](https://github.com/Joystream/joystream/issues/2409) (after the Olympia release): 
1. Benchmarking.
2. Refactoring the working group usage.
3. Moving the `ModuleAccount` to the common pallet.

Supercedes [this PR](https://github.com/Joystream/joystream/pull/2357).

shamil-gadelshin committed:
runtime: storage-v2: Initial commit

Pallet skeleton:
- crate creation
- basic types
- added events, errors, trait, storage stubs
- added signature of the upload extrinsic.

shamil-gadelshin committed:
runtime: storage-v2: Add create_storage_bucket extrinsic stub.

shamil-gadelshin committed:
runtime: storage-v2: Add testing infrastructure.

shamil-gadelshin committed:
runtime: storage-node: Add stubs for the storage provider workflow.

- new types
- added membership dependency
- temporary commented some types (dynamic bugs, distribution buckets)
- stubs: update_storage_buckets_for_static_bags, set_storage_operator_metadata, accept_pending_data_objects, accept_storage_bucket_invitation

shamil-gadelshin committed:
runtime: storage-v2: Add tests for create_storage_bucket().

shamil-gadelshin committed:
runtime: storage-v2: Add tests for create_storage_bucket().

shamil-gadelshin committed:
runtime: storage-v2: Add comments.

shamil-gadelshin committed:
runtime: storage-v2: Add benchmarking infrastructure.

shamil-gadelshin committed:
runtime: Add weights generation for the storage v2 pallet.

shamil-gadelshin committed:
runtime: storage-v2: Add accept_storage_bucket_invitation extrinsic.

shamil-gadelshin committed:
runtime: storage-v2: Start adding update_storage_buckets_for_static_bags extrinsic.

shamil-gadelshin committed:
runtime: storage-v2: Update tests.

shamil-gadelshin committed:
runtime: storage-v2: Implement upload extrinsic.

shamil-gadelshin committed:
runtime: storage-v2: Add set_storage_operator_metadata extrinsic.

shamil-gadelshin committed:
runtime: storage-v2: Add deletion_prize to uploading.

shamil-gadelshin committed:
runtime: storage-v2: Implement accept_pending_data_objects extrinsic.

shamil-gadelshin committed:
runtime: storage-v2: Add minor improvements.

- fix comments
- fix deletion_prize
- validate parameters for accepting pending data objects
- modify tests

shamil-gadelshin committed:
runtime: storage-v2: Add check for update_storage_buckets_for_static_bags

- check for non-existing storage bucket ids
- add comments

shamil-gadelshin committed:
runtime: storage-v2: Add working group static bags.

shamil-gadelshin committed:
runtime: storage-v2: Start adding dynamic bags.

- add dynamic bag ID and related types
- introduce bags type aliases

shamil-gadelshin committed:
runtime: storage-v2: Add BagManager helper struct.

shamil-gadelshin committed:
runtime: storage-v2: Refactor BagManager.

- introduce query and mutation abstractions.

shamil-gadelshin committed:
runtime: storage-v2: Add dynamic bag storage and tests.

shamil-gadelshin committed:
runtime: storage-v2: Modify update_storage_buckets_for_static_bags .

Make it work for both static and dynamic bags.

shamil-gadelshin committed:
runtime: storage-v2: Add cancel_storage_bucket_operator_invite extrinsic.

shamil-gadelshin committed:
runtime: storage-v2: Add invite_storage_bucket_operator extrinsic.

shamil-gadelshin committed:
runtime: storage-v2: Add uploading block.

shamil-gadelshin committed:
runtime: storage-v2: Add move_objects extrinsic.

shamil-gadelshin committed:
runtime: storage-v2: Add `move_data_objects` tests.

shamil-gadelshin committed:
runtime: storage-v2: Add `delete_data_objects` extrinsic.

shamil-gadelshin committed:
runtime: storage-v2: Add `update_storage_bucket_status` extrinsic.

shamil-gadelshin committed:
runtime: storage-v2: Add `update_blacklist` extrinsic.

shamil-gadelshin committed:
runtime: storage-v2: Remove `ipfs_content_id` from the data object.

shamil-gadelshin committed:
runtime: storage-v2: Add blacklist check on upload.

shamil-gadelshin committed:
runtime: storage-v2: Add blacklist limit checks.

shamil-gadelshin committed:
runtime: storage-v2: Change storage bucket flags.

Change accepting_new_data_objects to accepting_new_bag.

shamil-gadelshin committed:
runtime: storage-v2: Remove DAO types.

shamil-gadelshin committed:
runtime: storage-v2: Change StorageOperatorMetadata.

Change StorageOperatorMetadata to StorageBucketMetadata.

shamil-gadelshin committed:
runtime: storage-v2: Change `upload` extrinsic.

- convert `upload` extrinsic to the `upload_data_objects` public method
- rename `validate_upload_parameters` to the `can_upload_data_objects` and make it public.

shamil-gadelshin committed:
runtime: storage-v2: Change `move_data_objects` extrinsic.

Convert `move_data_objects` extrinsic to public method.

shamil-gadelshin committed:
runtime: storage-v2: Change `delete_data_objects` extrinsic.

Convert `delete_data_objects ` extrinsic to public method.

shamil-gadelshin committed:
runtime: storage-v2: Change storage bucket metadata storage.

Remove StorageBucketMetadata storage map and add `metadata` field to the `StorageBucket` struct.

shamil-gadelshin committed:
runtime: storage-v2: Start adding `voucher-feature`

shamil-gadelshin committed:
runtime: storage-v2: Add voucher usage on `upload_data_objects`

shamil-gadelshin committed:
runtime: storage-v2: Add voucher usage on `move_data_objects`

shamil-gadelshin committed:
runtime: storage-v2: Add voucher usage to `delete_data_objects`.

shamil-gadelshin committed:
runtime: storage-v2: Introduce `DataObjectStorage` trait.

shamil-gadelshin committed:
runtime: storage-v2: Move `BagManager` to a separate file.

shamil-gadelshin committed:
runtime: storage-v2: Add voucher usage to `update_storage_buckets_for_bags`.

shamil-gadelshin committed:
runtime: storage-v2: Add DistributionBucketId.

Add DistributionBucketId to the Trait and bags (both Static and Dynamic).

shamil-gadelshin committed:
runtime: storage-v2: Add `delete_dynamic_bags` method.

shamil-gadelshin committed:
runtime: storage-v2: Add `deletion_prize` handling to `delete_data_objects`

shamil-gadelshin committed:
runtime: storage-v2: Remove `number_of_pending_data_objects`.

Remove `number_of_pending_data_objects` field from the StorageBucket.

shamil-gadelshin committed:
runtime: storage-v2: Add `deletion_prize` usage to `move_data_objects`.

shamil-gadelshin committed:
runtime: storage-v2: Add event `DeletionPrizeChanged`.

shamil-gadelshin committed:
runtime: storage-v2: Add event `VoucherChanged`.

shamil-gadelshin committed:
runtime: storage-v2: Add `delete_storage_bucket` extrinsic.

shamil-gadelshin committed:
runtime: storage-v2: Modify `accept_pending_data_objects `

Add `storage_bucket_id`.

shamil-gadelshin committed:
runtime: storage-v2: Refactor `delete_dynamic_bag`.

Change inpurt parameters from multiple bags to a single one and rename.

shamil-gadelshin committed:
runtime: storage-v2: Refactor `delete_data_objects`

Change parameter types.

shamil-gadelshin committed:
runtime: storage-v2: Add `remove_storage_bucket_operator` extrinsic.

shamil-gadelshin committed:
runtime: storage-v2: Add data size fee for uploading.

- add new extrinsic `update_data_size_fee`
- add storage field `DataObjectPerMegabyteFee`
- refactor static bags (remove separate council and working groups storage fields)
- slash data size fee on uploading

shamil-gadelshin committed:
runtime: pallet_common: Refactor InputValidationLengthConstraint.

shamil-gadelshin committed:
runtime: storage-v2: Add `StorageBucketsPerBagValueConstraint` to Trait.

shamil-gadelshin committed:
runtime: storage-v2: Add `StorageBucketsPerBagLimit` parameter.

- add related const constraint `StorageBucketsPerBagValueConstraint`
- add changing extrinsic `update_storage_buckets_per_bag_limit`

shamil-gadelshin committed:
runtime: storage-v2: Modify `update_storage_buckets_for_bags` extrinsic.

- change parameters to affect single bag rather than multiple
- change parameters: split “replacing” bucket collection into two - “added ids” and “removed ids”
- add StorageBucketPerBagLimitExceeded check
- add `accepting_new_bags` check for buckets

shamil-gadelshin committed:
runtime: storage-v2: Add `set_storage_bucket_voucher_limits` extrinsic.

shamil-gadelshin committed:
runtime: storage-v2: Add minor changes.

- change comment
- change benchmarking

shamil-gadelshin committed:
runtime: storage-v2: Add `update_storage_buckets_voucher_max_limits` extrinsic.

shamil-gadelshin committed:
runtime: storage-v2: Refactor bag access methods.

shamil-gadelshin committed:
runtime: storage-v2: Add `DynamicBagType`

shamil-gadelshin committed:
runtime: storage-v2: Start adding `create_dynamic_bag` method.

shamil-gadelshin committed:
runtime: storage-v2: Add dynamic bag existence check.

shamil-gadelshin committed:
runtime: storage-v2: Add storage buckets picking for a dynamic bag.

shamil-gadelshin committed:
runtime: storage-v2: Fix randomness algorithm.

shamil-gadelshin committed:
runtime: storage-v2: Refactor randomness algorithm.

shamil-gadelshin committed:
runtime: storage-v2: Modify create_dynamic_bag method.

Add random storage bucket picking.

shamil-gadelshin committed:
runtime: storage-v2: Add changing to dynamic bag creation policies.

- add default values
- add changing `update_number_of_storage_buckets_in_dynamic_bag_creation_policy` extrinsic

shamil-gadelshin committed:
runtime: storage-v2: Add storage operator existence check.

shamil-gadelshin committed:
runtime: storage-v2: Modify `accept_pending_data_objects` extrinsic.

- add storage bucket relation check
- change input parameters

shamil-gadelshin committed:
runtime: storage-v2: Refactor the pallet.

- modify upload_data_objects() method - change voucher usage
- rename VoucherUpdateWithDeletionPrize to BagChangeInfo

shamil-gadelshin committed:
runtime: storage-v2: Change comments.

shamil-gadelshin committed:
runtime: Remove obsolete `storage` and `service-discovery` pallets.

shamil-gadelshin committed:
runtime: storage: Rename pallet.

shamil-gadelshin committed:
runtime: storage: Add module comment.

shamil-gadelshin committed:
runtime: storage: Refactor pallet.

- remove obsolete TODOs

shamil-gadelshin committed:
runtime: storage: Change BagId enum.

- rename StaticBag to Static
- rename DynamicBag to Dynamic

shamil-gadelshin committed:
runtime: storage: Fix incorrect ‘buckets-per-bag’ limit check.

Lezek123 commented on runtime-modules/storage/src/lib.rs:
```diff
@@ -1,20 +1,2373 @@
+        /// Params
+        /// - data objects IDs
+        /// - initial uploading parameters
+        DataObjectdUploaded(Vec<DataObjectId>, UploadParameters),
```
```suggestion
        DataObjectsUploaded(Vec<DataObjectId>, UploadParameters),
```

shamil-gadelshin replied:
Will fix it. Thank for finding.

Lezek123 commented on runtime-modules/storage/src/lib.rs:
```diff
@@ -1,20 +1,2373 @@
+            operation_type,
+        );
+
+        Self::deposit_event(RawEvent::DataObjectdUploaded(
```
```suggestion
        Self::deposit_event(RawEvent::DataObjectsUploaded(
```

shamil-gadelshin replied:
Will fix it. Thank for finding.

Lezek123 commented on runtime-modules/storage/src/tests/mod.rs:
```diff
@@ -1,6 +1,3551 @@
+            DataObjectDeletionPrize::get()
+        );
+
+        EventFixture::assert_last_crate_event(RawEvent::DataObjectdUploaded(
```
```suggestion
        EventFixture::assert_last_crate_event(RawEvent::DataObjectsUploaded(
```

Lezek123 reviewed (changes requested):
Just noticed a small typo

bedeho commented on runtime-modules/storage/src/lib.rs:
```diff
@@ -1,20 +1,2373 @@
+
+            let bucket = Self::ensure_storage_bucket_exists(&storage_bucket_id)?;
+
+            Self::ensure_bucket_missing_invitation_status(&bucket)?;
```
Should we _require_ that there is not even a pending invitation? Perhaps a bit overly strict for no clear benefit.

shamil-gadelshin replied:
There are at least two possible states: 1. the same worker was invited, 2. another worker was invited.

Do we want to have the same outcome of the extrinsic with these states?

Should we silently replace the current invite (or maybe with a different event or even with a different event depending on the state)? 

Requiring to cancel the previous invite is a straightforward approach. However, this also increases the number of extrinsic calls (+1 for canceling). If you prefer to simplify usability - I recommend default invite replacement with the same event. Please, let me know if you still think we need to replace the invite.

bedeho replied:
> Please, let me know if you still think we need to replace the invite.

We are deleting the bucket here, so its not a matter of replacing, its just a matter of entirely dropping the bucket, which by implication drops storage state which reflects the invite.

shamil-gadelshin replied:
Your comment relates to the `invite_storage_bucket_operator` extrinsic judging by line number 1357. But it looks you mean the `delete_storage_bucket` extrinsic which has the same `Self::ensure_bucket_missing_invitation_status(&bucket)?;` check. Yes, it is safe to delete the check here - will do.

shamil-gadelshin replied:
Fixed in https://github.com/Joystream/joystream/pull/2407/commits/1ee491524e7a22cfbb8d2eac2fa4ecfce002b44e

bedeho commented on runtime-modules/storage/src/lib.rs:
```diff
@@ -1,20 +1,2373 @@
+            bag_change.total_deletion_prize,
+        )?;
+
+        Self::slash_data_size_fee(
```
Given that the lead can unilaterally set the fee to any value they want, this could allow them to steal/burn an arbitrary quantity from a user who is unwittingly holding large quantities of funds in the account they use to fund these funds with. This can even be done after a user has sent a transaction if timed well. In particular if we are going to divert part of the fee to the storage provider confirming the upload, then it could become very attractive.

Should we remedy this by advising all users of this extrinsic to commit to the fee they are offering in the extrinsic that ultimately calls this method `upload_data_objects` here?

bedeho replied:
I meant to mention this earlier, but I think its probably prudent to crediting this fee to the eventual provider that correctly confirms the upload, in order to provide an incentive for this to be done correctly. This will require inclusion of some auth step in the upload process, so that only the genuine uploader can confirm. But lets do this in a later PR.

shamil-gadelshin replied:
Ok. I return `DataSizeFeeChanged` error on fee change now.

bedeho replied:
Fixed in https://github.com/Joystream/joystream/pull/2407/commits/65a81779c274383c08bc5900622e39e0a0f90ea1

bedeho commented on runtime-modules/storage/src/lib.rs:
```diff
@@ -1,20 +1,2373 @@
+        type Error = Error<T>;
+
+        /// Exports const - max allowed storage bucket number.
+        const MaxStorageBucketNumber: u64 = T::MaxStorageBucketNumber::get();
```
I wonder if we need to cap this? buckets are just stored in a standalone map, so it's not clear if there is a big upside here, also, the number may need to get really big, in which case having a hard constant could end up being a bad problem.

shamil-gadelshin replied:
Ok. We don't have any iterations with `StorageBucketById` so it is safe to delete now.

bedeho commented on runtime-modules/storage/src/lib.rs:
```diff
@@ -1,20 +1,2373 @@
+            }
+
+            Self::deposit_event(
+                RawEvent::StorageBucketsUpdatedForBag(bag_id, add_buckets, remove_buckets)
```
I think it would perhaps be better to only have a single event per extrinsic. I know we do not do this consistently prior, but it seems more clean to have one singular event carrying all side-effects of a given extrinsic. Currently, there will be up to two `VoucherChanged` events in this extrinsic prior to this final one.

shamil-gadelshin replied:
I send events inside small methods closer to the storage changing code. It is very easy to forget about such events in upstream code. I started to do like that after you requested to reveal side effects in the bounty pallet. There is another such event in the storage pallet - `DeletionPrizeChanged` and there can be more than two or three such events (there is a loop inside the voucher changing function). Please, let me know if you think that forgetting about side effects is a minor issue and I will remove such hidden events in favor of more complicated upper-level events.

bedeho replied:
I understand what you mean, it's a bit hard for me to judge tbh. I see the possible problems of both approaches. I guess we can just continue as is for now, and see what ends up happening upstream.

shamil-gadelshin replied:
Ok.

bedeho commented on runtime-modules/storage/src/lib.rs:
```diff
@@ -1,20 +1,2373 @@
+
+        /// Sets storage bucket voucher limits.
+        #[weight = 10_000_000] // TODO: adjust weight
+        pub fn set_storage_bucket_voucher_limits(
```
Only lead can set this.

shamil-gadelshin replied:
Don't workers control their storage capacity (set_storage_bucket_voucher_limits), and a leader should only control the maximum possible capacity (update_storage_buckets_voucher_max_limits), do they? Workers seem to know more about their system's capabilities.

bedeho replied:
No the capacity is in fact what they are paid for, not just actual utilization, so its not up to their own discretion.

shamil-gadelshin replied:
Ok.

shamil-gadelshin replied:
Fixed in https://github.com/Joystream/joystream/pull/2407/commits/1ee491524e7a22cfbb8d2eac2fa4ecfce002b44e

bedeho commented on runtime-modules/storage/src/lib.rs:
```diff
@@ -1,20 +1,2373 @@
+
+        /// Update whether new bags are being accepted for storage.
+        #[weight = 10_000_000] // TODO: adjust weight
+        pub fn update_storage_bucket_status(
```
Only lead can set this.

shamil-gadelshin replied:
Ok.

bedeho commented on runtime-modules/storage/src/lib.rs:
```diff
@@ -1,20 +1,2373 @@
+            Error::<T>::DataObjectsPerBagLimitExceeded
+        );
+
+        let mut bag_change = BagChangeInfo::default();
```
Can we do a pure version of this, no `mut`, should be easy.

shamil-gadelshin replied:
Ok.

bedeho commented on runtime-modules/storage/src/lib.rs:
```diff
@@ -1,20 +1,2373 @@
+            Error::<T>::NoObjectsOnUpload
+        );
+
+        BagManager::<T>::ensure_bag_exists(&params.bag_id)?;
```
This looks a bit fishy to me, as you are not returning an actual bag. This must mean that in many code paths you are going to be looking up that bag again, possibly with a fallible lookup, or a blind mutation, where the latter could risk updating an invalid (default constructed) instance.

shamil-gadelshin replied:
Ok. BagManager has a state now (bag_id), and I combined its construction with an existence check.

bedeho replied:
Does this really fix the problem? If you return a `BagManager`, you have lost the static guarantee that the bag exists, which was the main issue to begin with. Seems to me that if the way a manager is used is that its always kept around after a bag has been shown to exist, then you need to keep a copy in the manager, however, you are mostly not using these manager instances in a statefule way, so I am not sure that really makes sense. Perhaps the semantics of what this manager represents i a bit unclear, let me know what you think, and just make the decision you think is best.

shamil-gadelshin replied:
The `BagManager` is an abstraction over the static and dynamic bags storage and types. 

It handles different structs for `DynamicBag` and `StaticBag` and different storage maps for static and dynamic bags. Currently, there are only two types of storage for bags. One combined storage map for the static bags (council and working groups) and one for all dynamic bags(Member and Channel). But we could fall back to the initial storage design with one storage map per bag type without changing the code that uses the BagManager.

For example, instead of 

`BagManager::<T>::ensure_bag_exists(&bag_id)?;`

we could write

```
match bag_id{
    Static(static_bag_id) => ensure_static_bag_exists(static_bag_id)
    Dynamic(dynamic_bag_id) => ensure_dynamic_bag_exists(dynamic_bag_id)
}?;
```

The code for getting the bag, manipulating with its content and saving back will be the same for all storages. So, I absracted it with the `Bagmanager`.

However, this all goes away if we drop the `deletion prize` field in the dynamic bag struct (your another comment). Because, it makes the `StaticBag` struct identical with the `DynamicBag` struct. And I can even create a single storage map `BagId => Bag` and remove a huge part of `BagManager` or drop it completely.

bedeho replied:
1. I understand, that is indeed practical, I think my point in that particular case would be to return some sum type across the two bag types, so that the ensure abstraction is similar to how we typically do it. But I dont know if that causes other problems later.
2. I suggested a way to approach having just a single bag type here https://github.com/Joystream/joystream/pull/2407#discussion_r676404403, if that sounds advisable, then lets do that. However, let us not put static bags into the dynamic map, as that makes the representations much less safe, as static bags can be deleted and so on. Let just use the same approach as now, but with only one bag type.

bedeho commented on runtime-modules/storage/src/lib.rs:
```diff
@@ -1,20 +1,2373 @@
+
+// Helper struct for the dynamic bag changing.
+#[derive(Clone, PartialEq, Eq, Debug, Copy, Default)]
+struct BagChangeInfo<Balance> {
```
Can we call this `BagSideffects` or something, `Info` kills me 🗡️

shamil-gadelshin replied:
Renamed to `BagUpdate`

bedeho commented on runtime-modules/storage/src/lib.rs:
```diff
@@ -1,20 +1,2373 @@
+    pub distributed_by: BTreeSet<DistributionBucketId>,
+
+    /// Dynamic bag deletion prize.
+    pub deletion_prize: Balance,
```
What was the reason for having a separate prize here that also incorporates the size of `objects`, when each `DataObject` has its own prize to begin with? I think there was a reason, but I don't recall now.

shamil-gadelshin replied:
It spares us from iteration through all data objects (could be expensive).

bedeho replied:
1) Does that mean it is redundant, it is the sum of all the data objects?
2) When the objects are inlined, then its just iterating in memory, that is going to be super fast, and there is even going to be a length cap, so I don't think that is a real issue.
3) I believe we are still going for non-inlined dataobjects, at least last we spoke I argued pattionatley for this 🙏 📿 , in which case deletion of non-empty bags is not going to be a thing anyway.

shamil-gadelshin replied:
1. 'Redundant' in the meaning of 'cached' or 'denormalized'.
2. I'm not sure 'super fast' would be still the case for large bags (starting with hundreds of thousands data objects). However, a large bag deletion looks like a very rare operation. And we're committing a primordial sin of preliminary optimization. So, let's try to drop the field until stress tests show us the light of wisdom. There are additional implications - see another comment about the `BagManager`.
3. Please, remind me: didn't we agree to prevent deleting non-empty bags forcing the leader to manually delete possibly thousands of objects considering it as a rare operation, and drop the inlined data objects map in favor of a double map `BagId => DataObjectId => DataObject`. If we don't prevent non-empty bag deletion - we need to iterate through the double-map prefix by BagId. The only iteration would be also on deleting the dynamic bag.

bedeho replied:
1. I reviewed this again, and I think the rationale for having this is distinct from the data objects. Remember, even if you have 0 videos in your channel, for example, we still need to give you an incentive to throw out the channel itself. That is what this prize does. This means it is needed, even if we have non-inlined objects. However, I do wonder, should we have the calling code manage this incentive itself, thereby dropping the distinction between dynamic and static bags in this module? In this alternative approach, a channel would have its own storage prize in the content directory, and that prize would be calibrated by that code to give sufficient incentive to cleanup both that channel there, and the dynamic bag in the storage system. Let me know what you think.
3. Yes I think this is what we agreed, this is what  I was trying to say. Note that the storage lead is not the deleter, it has to be the owner of the bag.

bedeho commented on runtime-modules/storage/src/lib.rs:
```diff
@@ -1,20 +1,2373 @@
+
+// Public methods
+impl<T: Trait> DataObjectStorage<T> for Module<T> {
+    fn can_upload_data_objects(params: &UploadParameters<T>) -> DispatchResult {
```
Would it be possible to make all these action extrinsics infallible provided the user provides a `bag_change` variable which they received from a positive call to  `can_*`.

shamil-gadelshin replied:
Yes, it is possible. However, such an interface has a couple of downsides:
1. It would disclose several internal helper abstractions like BagChange, VoucherUpdate, BagManager, etc. It would complicate the pallet API with no obvious benefit.

2. There will be a way to call the methods with manually constructed data bypassing the actual checks.

3. It arguably looks like a contradiction with your other comment about 
`BagManager::<T>::ensure_bag_exists(&params.bag_id)?;`

```
.... 
This must mean that in many code paths you are going to ...<use this abstraction again> ...
...possibly with a fallible lookup, or a blind mutation, where the latter could risk updating an invalid (default constructed) instance.
```

bedeho replied:
I agree.

bedeho reviewed (changes requested):
Awesome job! I did a decent level review, not as rigorous as I could, have, but still useful I think.

Reading this made me question whether we really should think of the distribution bucket operator needs a new id space. Ultimately, for each combination of worker & distribution bucket you have at most one relationship at any given time, as the same worker cannot operate multiple times on the same bucket at the same time. Off-chain one may want to identify distinct occasions where the same worker has a role in the same bucket over time, for example if they are added and removed multiple times, but the runtime does not need to care about this distinction.

shamil-gadelshin committed:
runtime: storage: Remove a constant.

- remove `MaxStorageBucketNumber` constant

shamil-gadelshin committed:
runtime: storage: Change BagManager helper.

shamil-gadelshin committed:
runtime: storage: Rename BagChangeInfo to BagUpdate

shamil-gadelshin committed:
runtime: storage: Refactor ‘mut’ variables.

shamil-gadelshin committed:
runtime: storage: Change extrinsic.

- change `update_storage_bucket_status`:
replace worker origin check with leader origin check.

shamil-gadelshin committed:
runtime: storage: Add parameter for public method.

- add `expected_data_size_fee` for the `upload_data_objects` public method.

shamil-gadelshin committed:
runtime: storage: Fixing review comments.

- remove a check on deleting the storage bucket
- change required permissions for setting the storage bucket voucher.

shamil-gadelshin committed:
runtime: storage: Remove metadata from the state.

shamil-gadelshin committed:
runtime: storage: Simplify bag storage.

- Merge static and dynamic structs and storage.
- Remove the BagManager
- Add deletion prize on the dynamic bag creation.

shamil-gadelshin committed:
runtime: storage: Move data objects to the double map.

shamil-gadelshin committed:
runtime: storage: Add `BagObjectsChanged` event.

shamil-gadelshin commented:
Changes summary:
- replaced DynamicBag and StaticBag structs for a single Bag struct. Added optional deletion prize for a bag
- removed BagManager abstraction
- introduced deletion prize withdrawal and deposit on dynamic bag creation/deletion.
- replaced inlined data objects map with storage double map BagId => DataObjectId => DataObject. I chose the double map over the simple map to avoid introducing an additional map with bag-to-object relations - the double map has it by design.
- introduced total objects size and number counters for bags instead of ad-hoc calculation using inlined data objects map and their changing event.

bedeho reviewed (changes requested):
Excellent, only thing I would have us change is to not use the name `BagObject` and `BagDataObjectsById` , and just use `DataObject`.

shamil-gadelshin committed:
runtime: storage: Rename BagDataObjectsById map.

BagDataObjectsById => DataObjectsById
